internal/devpkg/package.go

```diff
@@ -462,8 +470,34 @@ func (p *Package) IsInBinaryCache() (bool, error) {
 	}
 
 	// Check if the user's system's info is present in the lockfile
-	_, ok := entry.Systems[userSystem]
-	return ok, nil
+	sysInfo, ok := entry.Systems[userSystem]
+	if !ok {
+		return false, nil
+	}
+
+	// Check if the narinfo is present in the binary cache
+	if exists, ok := isNarInfoInCache[p.Raw]; ok {
+		fmt.Printf("narInfo cache hit: %v\n", exists)
+		return exists, nil
+	}
+
+	httpClient := &http.Client{
+		Timeout: time.Second * 5,
+	}
+	pathParts := newStorePathParts(sysInfo.StorePath)
+	reqURL := BinaryCache + "/" + pathParts.hash + ".narinfo"
+	res, err := httpClient.Head(reqURL)
```

Comment:
Adding an HTTP request in `IsInBinaryCache` will likely introduce N+1s. I generally try to avoid network requests in struct methods for this reason. 

In practice this means we will make 1 sequential network call per package any time local.lock is out of date. A few alternatives:

* Only check on failure. If nix profile install or print-dev-env fail, then check.
* Make http calls in parallel. This may break if cache has any sort of rate limit (though if it's just S3 I doubt it).
* If cache is never evicted, we can check on installation instead. If it's missing from cache, mark as missing in lockfile.

Reply:
Speaking with @gcurtis it appears there's some likelihood that the nix maintainers start evicting old packages from the cache.

So, given that, option 2 seems the better path to pursue. A rate-limit doesn't seem to be a big concern: nix itself does a lot of requests to the cache, and when populating the search index @gcurtis reports doing many requests to the cache in parallel without getting rate-limited.

option 1 of checking upon failure I am ruling out because it seems quite hard to implement.